friends/src/App.js

```diff
@@ -0,0 +1,126 @@
+import Friend from './components/FriendsComponents/Friend'
+import Header from './components/Header'
+import Form from './components/SharedComponents/Form'
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      friends: []
+    };
+    console.log(`App props are: `, this.props)
+  }
+
+  updateFriends = data => {
+    this.setState({ friends: data },
+      () => console.log(`updateFriends invoked state is: `, this.state)
+    )
+  }
+
+  deleteFriend = id => {
+    console.log("Friend is being deleted")
+    axios
+      .delete(`http://localhost:5000/friends/${id}`)
+      .then(response => {
+        // Update main app state
+        this.updateFriends(response.data)
+        this.props.history.push('/')
+      })
+      .catch(err => console.log(err))
+  };
+
+  componentDidMount() {
+    axios
+      .get('http://localhost:5000/friends')
+      .then(response => {
+        this.setState({ friends: response.data },
+          () => console.log(`CDM invoked state is: `, this.state))
+      })
+      .catch(error => console.log(`Data retrieval error: ${error}`))
+  }
+
+  render() {
+    console.log(`render state is: `, this.state)
+    return (
+      <AppContainer>
+        <Header />
+        {/* Add routes */}
+        <Route
+          exact path='/'
+          render={props => 
+            <FriendsList 
+              {...props} 
+              {...this.state}
+            />}
+        />
+        <Route 
+            path='/add'
+            render={props => 
+              <Form 
+                {...props}
+                updateFriends={this.updateFriends}
+                add
+              />}
+        />
+        <Route
+            exact
+            path='/update'
+            render={props => 
+              <Form 
+                {...props}
+                {...this.state}
+                updateFriends={this.updateFriends}
+                update
+              />}
+        />
+        <Route 
+            path='/delete'
+            render={props => 
+              <Form 
+                {...props}
+                updateFriends={this.updateFriends}
+                deleteFriend={this.deleteFriend}
+                delete
+              />}
+        />
+        {this.state.friends.map(friend => (
+          <Route
+            key={friend.id}
+            path={`/friends/${friend.id}`}
+            render={props => 
+              <Friend 
+                {...props} 
+                friend={friend} 
+                deleteFriend={this.deleteFriend}
+              />
+            }
+          />
+        ))}
+        {this.state.friends.map(friend => (
+          <Route
+            key={friend.id}
+            path={`/update/${friend.id}`}
+            render={props =>
+              <Form
+                {...props}
+                friend={friend}
+                id={friend.id}
+                updateFriends={this.updateFriends}
+                update
```

Comment:
Very nice work with all these routes. I appreciate the consistent & readable formatting and adherence to naming conventions.